debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/TypeRegistry.java

```diff
@@ -319,4 +289,265 @@ public int ltreeArrayOid() {
     public static String normalizeTypeName(String typeName) {
         return LONG_TYPE_NAMES.getOrDefault(typeName, typeName);
     }
+
+    /**
+     * Prime the {@link TypeRegistry} with all existing database types
+     */
+    private void prime() {
+        Connection pgConnection = null;
+        try {
+            pgConnection = connection.connection();
+
+            final TypeInfo typeInfo = ((BaseConnection) pgConnection).getTypeInfo();
+            final SqlTypeMapper sqlTypeMapper = new SqlTypeMapper(pgConnection, typeInfo);
+
+            try (final Statement statement = pgConnection.createStatement()) {
+                // Read non-array types
+                try (final ResultSet rs = statement.executeQuery(SQL_NON_ARRAY_TYPES)) {
+                    while (rs.next()) {
+                        // Coerce long to int so large unsigned values are represented as signed
+                        // Same technique is used in TypeInfoCache
+                        final int oid = (int) rs.getLong("oid");
+                        final int baseOid = (int) rs.getLong("baseoid");
+                        final int modifiers = (int) rs.getLong("modifiers");
+                        String typeName = rs.getString("name");
+
+                        // If a base type reference exists for this type, lookup base type.
+                        PostgresType baseType = null;
+                        if (baseOid != 0) {
+                            baseType = get(baseOid);
+                        }
+
+                        addType(new PostgresType(
+                                typeName,
+                                oid,
+                                sqlTypeMapper.getSqlType(typeName),
+                                modifiers,
+                                typeInfo,
+                                baseType,
+                                null));
+                    }
+                }
+
+                // Read array types
+                try (final ResultSet rs = statement.executeQuery(SQL_ARRAY_TYPES)) {
+                    while (rs.next()) {
+                        // int2vector and oidvector will not be treated as arrays
+                        final int oid = (int) rs.getLong("oid");
+                        final int baseOid = (int) rs.getLong("baseoid");
+                        final int modifiers = (int) rs.getLong("modifiers");
+                        String typeName = rs.getString("name");
+
+                        // If a base type reference exists for this type, lookup base type
+                        PostgresType baseType = null;
+                        if (baseOid != 0) {
+                            baseType = get(baseOid);
```

Comment:
If `baseOid` is not yet in the registry then `resolveUnknownType` is triggered which executes database query.
Would not this be better done in that way that firs all types are loaded and then the tree-like structure (`baseOid`) is built from it?

Reply:
Possibly; my assumption was that a non-array type would only extend a non-array type and that the `ORDER BY` specified in the query would manage this but perhaps it may not in some corner cases in hindsight.  Good call.